titan-compiler/coder.lua

```diff
@@ -476,34 +476,60 @@ end
 -- @records
 --
 
--- Compute the layout for the gc fields stored in the UValue array.
--- We use a C struct to compute the layout of fields stored in mem part.
+-- Compute the record layout. We store the gc fields in UValue array and
+-- primitive fields in the mem part.
 local function rec_compute_layout(rec)
-    local index = 0
+    local gc_index = 0
+    local prim_index = 0
     local pos = {}
     for _, field in ipairs(rec.field_decls) do
         local typ = rec._field_types[field.name]
         if types.is_gc(typ) then
-            pos[field.name] = index
-            index = index + 1
+            pos[field.name] = gc_index
+            gc_index = gc_index + 1
+        else
+            pos[field.name] = prim_index
+            prim_index = prim_index + 1
         end
     end
-    return { pos = pos, size = index }
+    return { pos = pos, gc_size = gc_index, prim_size = prim_index }
 end
 
 local function rec_struct_name(rec)
     return string.format("struct record_%s", rec.name)
 end
 
+local function rec_field_name(rec, field_name)
+    return string.format("field_%d", rec._layout.pos[field_name])
+end
+
+local function rec_gc_size(rec)
+    return rec._layout.gc_size
+end
+
+local function rec_mem_size(rec)
+    local layout = rec._layout
+    if layout.prim_size == 0 then
+        return 0
+    else
+        return string.format("sizeof(%s)", rec_struct_name(rec))
+    end
+end
+
 local function rec_declare_struct(rec)
+    if rec._layout.prim_size == 0 then return "" end
```

Comment:
Please add a comment clarifying that empty structs are not allowed by the C standard and that they are a GCC extension. (I think here is the best spot but anywhere would be fine)

Reply:
Will do.